File: bumperbot_vision/bumperbot_vision/undistort_image.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import Image, CameraInfo
from cv_bridge import CvBridge
from image_geometry import PinholeCameraModel
import message_filters
import numpy as np

class UndistortImage(Node):
    def __init__(self):
        super().__init__('undistort_image')

        # Declare parameters
        self.declare_parameter('camera_name', 'left_camera')
        self.camera_name_ = self.get_parameter('camera_name').get_parameter_value().string_value

        # Initialize helpers
        self.bridge = CvBridge()
        self.model = PinholeCameraModel()

        image_sub = message_filters.Subscriber(
            self, 
            Image, 
            f'{self.camera_name_}/image_raw', 
            qos_profile=qos_profile_sensor_data
        )
        
        info_sub = message_filters.Subscriber(
            self, 
            CameraInfo, 
            f'{self.camera_name_}/camera_info', 
            qos_profile=qos_profile_sensor_data
        )

        self.time_sync = message_filters.ApproximateTimeSynchronizer(
            [image_sub, info_sub], 
            queue_size=10, 
            slop=0.1
        )
        self.time_sync.registerCallback(self.image_callback)

        self.undistort_image_pub_ = self.create_publisher(
            Image, 
            f'{self.camera_name_}/image_rect', 
            qos_profile_sensor_data
        )

        self.get_logger().info(f'Undistort node started for {self.camera_name_}')

    def image_callback(self, image_msg, info_msg):
        cv_image = self.bridge.imgmsg_to_cv2(image_msg, desired_encoding="passthrough")
        self.model.from_camera_info(info_msg)

        image_undistort = np.zeros_like(cv_image)
        self.model.rectify_image(cv_image, image_undistort)
        undistort_msg = self.bridge.cv2_to_imgmsg(image_undistort, encoding=image_msg.encoding)
        undistort_msg.header = image_msg.header
        self.undistort_image_pub_.publish(undistort_msg)


def main(args=None):
    rclpy.init(args=args)
    node = UndistortImage()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: bumperbot_vision/src/undistort_image.cpp
#include "bumperbot_vision/undistort_image.hpp"

#include "cv_bridge/cv_bridge.hpp"

#include "opencv2/core.hpp"
#include "opencv2/opencv.hpp"

namespace bumperbot_vision
{
UndistortImage::UndistortImage(const std::string & name) : Node(name)
{
  approximate_sync_ = std::make_shared<ApproximateSync>(
    ApproximatePolicy(10), image_sub_, info_sub_);
  approximate_sync_->registerCallback(
    std::bind(&UndistortImage::imageCallback, this, _1, _2));
  image_transport::TransportHints hints{this};
  const auto sensor_data_qos = rclcpp::SensorDataQoS().get_rmw_qos_profile();

  // Declare parameters
  camera_name_ = declare_parameter("camera_name", "left_camera");

  image_sub_.subscribe(
    this, camera_name_ + "/image_raw", hints.getTransport(), sensor_data_qos);
  info_sub_.subscribe(this, camera_name_ + "/camera_info", sensor_data_qos);

  undistort_image_pub_ =
    image_transport::create_publisher(this, camera_name_ + "/image_rect", rmw_qos_profile_sensor_data);
}

void UndistortImage::imageCallback(
  const sensor_msgs::msg::Image::ConstSharedPtr & image_msg,
  const sensor_msgs::msg::CameraInfo::ConstSharedPtr & info_msg)
{
  cv::Mat image = cv_bridge::toCvCopy(image_msg, image_msg->encoding)->image;
  int rows = image.rows, cols = image.cols;
  cv::Mat image_undistort;

  model_.fromCameraInfo(info_msg);
  model_.rectifyImage(image, image_undistort, 1);

  auto undistort_msg =
    cv_bridge::CvImage(image_msg->header, image_msg->encoding, image_undistort).toImageMsg();
  undistort_msg->header = image_msg->header;
  undistort_image_pub_.publish(undistort_msg);
}
}  // namespace bumperbot_vision

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<bumperbot_vision::UndistortImage>("undistort_image"));
  rclcpp::shutdown();
  return 0;
}